### PR TITLE
[FLINK-14747] Fix performance regression with latest changes to Mailbox

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

* Fixes performance regression introduced with FLINK-14304

## Brief change log

* Reintroduce volatile variable to avoid lock acquisition on empty queue.

## Verifying this change

*(Please pick either of the following options)*

* Correctness covered by existing tests.
* Performance visible in benchmarks.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (yes)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxImpl.java [hunk 1]
```diff
@@ -80,6 +80,11 @@
 	 */
 	private final Deque<Mail> batch = new ArrayDeque<>();
 
+	/**
+	 * Performance optimization where hasNewMail == !queue.isEmpty().
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxImpl.java]: comment that it doesn't take the `batch` into account?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxImpl.java [hunk 2]
```diff
@@ -149,13 +144,17 @@ public boolean hasMail() {
 	@Override
 	public boolean createBatch() {
 		checkIsMailboxThread();
+		if (!hasNewMail) {
+			return !batch.isEmpty();
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxImpl.java]: Currently, we do, but it doesn't make any sense from the current usage pattern. So we could just go with the check state.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxImpl.java [hunk 3]
```diff
@@ -165,7 +164,8 @@ public boolean createBatch() {
 	@Override
 	public Optional<Mail> tryTakeFromBatch() {
 		checkIsMailboxThread();
-		return Optional.ofNullable(takeOrNull(batch, MIN_PRIORITY));
+		checkTakeStateConditions();
+		return Optional.ofNullable(batch.pollFirst());
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxImpl.java]: This is a performance improvement as it avoids creating an iterator.

[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxImpl.java]: Ok, I get it know. Maybe move it to a separate commit with an appropriate commit message explaining this?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxImpl.java]: I have no hard feelings either way :)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxImpl.java [hunk 4]
```diff
@@ -205,12 +207,16 @@ public void putFirst(@Nonnull Mail mail) {
 
 	@Nullable
 	private Mail takeOrNull(Deque<Mail> queue, int priority) {
-		checkTakeStateConditions();
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxImpl.java]: bump?

[the remark this answers is not in the document]
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxImpl.java]: It's related to the correct behavior of `take` and `tryTake`. Internally, they know invoke that function `takeOrNull` twice, which would result in two state checks. By moving that to the callee, it's now a mini-tad faster ;).

[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxImpl.java]: Ok, I see now :)

**Pull request merged**